cpp/include/cudf/io/detail/data_casting.cuh

```diff
@@ -350,36 +355,44 @@ std::unique_ptr<column> parse_data(str_tuple_it str_tuples,
                                    size_type col_size,
                                    data_type col_type,
                                    B&& null_mask,
+                                   size_type null_count,
                                    cudf::io::parse_options_view const& options,
                                    rmm::cuda_stream_view stream,
                                    rmm::mr::device_memory_resource* mr)
 {
   CUDF_FUNC_RANGE();
+
+  auto d_null_count    = rmm::device_scalar<size_type>(null_count, stream);
+  auto null_count_data = d_null_count.data();
+
   if (col_type == cudf::data_type{cudf::type_id::STRING}) {
-    // this utility calls the functor to build the offsets and chars columns
+    // this utility calls the functor to build the offsets and chars columns;
+    // the bitmask and null-count may be updated by parse failures
     auto [offsets, chars] = cudf::strings::detail::make_strings_children(
       string_parse<decltype(str_tuples)>{
-        str_tuples, static_cast<bitmask_type*>(null_mask.data()), options},
+        str_tuples, static_cast<bitmask_type*>(null_mask.data()), null_count_data, options},
```

Comment:
Minor nit: for consistency with the mask could just use `d_null_count.data()` directly here rather than defining a temp variable.

Reply:
Ah I see why you chose this approach, the variable is also used in the non-string code path. I'm fine with either approach.